### DISCUSSION TITLE
How to get standard cell data (size, pin count, logic function) from the gates of a design?

### DISCUSSION THREAD

**Discussion by gudeh** [The-OpenROAD-Project/OpenROAD-flow-scripts#1517] [Q&A]
Hi everyone! I would like to build a TCL script to get the standard cell data of each logic gate in the design. I wonder if it is even possible to extract these from ORFS TCL commands.

I was trying to at least print the names of the logic gates, it runs without errors, although it seems that "get_lib_cells" is always empty. I added this code to the end of "gui_final" script:
```
    set cell_names [get_lib_cells]
    set csv_file "cell_names.csv"
    set csv_handle [open $csv_file "w"]
    foreach cell $cell_names {
        puts $csv_handle $cell
    }
    close $csv_handle
```

**Comment by maliberty** [comment]
You need an argument, eg
```
>>> llength [get_lib_cells *]
135
````

**Reply by maliberty** [reply]
(I think it is a bug that * is required but I'll take that up separately)

**Reply by gudeh** [reply]
Indeed that worked. Although, I was able to get the name with get_name or get_full_name. The data is actually from the liberty file, although I would like to have the data based on the instantiated gates in the design. Can I do that?

```
set csv_file [open "out.csv" w]
puts $csv_file "Gate Name,Size"
foreach cell [get_lib_cells *] {
    set cell_fname [get_full_name $cell]
    set cell_name [get_name $cell]
    puts $csv_file "$cell_fname,$cell_name"
}
close $csv_file
```

I understand this issue [#1614](https://github.com/The-OpenROAD-Project/OpenROAD/issues/1614) also requested something similar.

**Reply by gudeh** [reply]
oh! I realized I wanted `get_cells`. Unfortunately the documentation does not mention a lot about it. What other type of data can I get other than the name with `get_name`?

**Reply by maliberty** [reply]
In sdc instances are called cells. What information are you trying to query? The timing info will be available through sdc but not physical info

**Reply by gudeh** [reply]
The cell area, number of input and output pins, and the string logic function, if possible.

**Reply by gudeh** [reply]
Well, obviously this can be achieved by reading the liberty file myself. I did with python and it is working fine. 

I wanted to do this via OpenROAD because I assume it would be safer and also to learn a little more about the tool. Unfortunately the documentation does not include a lot about TCL commands.

I also attempted to do so with Yosys RTLIL, but I was not successful. It seems to not be possible to get standard cell data from RTLIL.

Either way, I am just replying here to wrap up this question.

**Reply by maliberty** [reply]
Have you looked at https://github.com/The-OpenROAD-Project/OpenSTA/blob/master/doc/OpenSTA.pdf ?

**Reply by maliberty** [reply]
I don't think the function is exposed but the others are easy to get from the odb APIs and possible from sta.
